File: apps/desktop/src/lib/subagent-display.ts
import type { ConversationMessageSnapshot, PendingAssistantAux } from "../types.js";

/** Parent wrap-up after subagent — normal assistant body, not runtime status.
 * The CJK patterns intentionally match model-generated Chinese wrap-up text; models may
 * still answer in Chinese, so this detector stays CJK-aware by design. */
function isParentSubagentCompletionSurfaceText(text: string): boolean {
  return /子智能体已完成|输出如下/u.test(text);
}

/** Host-drawn spinner frame prefix on live pending aux status lines. */
const SUBAGENT_SPINNER_PREFIX = /^[|/\\-]\s+/;

/** Progress tail after `title:` on the subagent status line (streaming English fragments included). */
const SUBAGENT_STATUS_TAIL_PREFIX = /^(The|Sub|Sp|Thinking|Compressing|Running|Awaiting)\b/u;

/** Colon is part of an emoticon (e.g. `:)`), not a `label: status` separator. */
function isEmoticonColon(text: string, colonIdx: number): boolean {
  const next = text[colonIdx + 1];
  return next !== undefined && /[)D(P/\\\]oO0-3]/.test(next);
}

function lastStatusColonIndex(text: string): number {
  let colonIdx = Math.max(text.lastIndexOf(":"), text.lastIndexOf("："));
  while (colonIdx > 0 && isEmoticonColon(text, colonIdx)) {
    const prevAscii = text.lastIndexOf(":", colonIdx - 1);
    const prevFull = text.lastIndexOf("：", colonIdx - 1);
    const prev = Math.max(prevAscii, prevFull);
    if (prev <= 0) {
      return -1;
    }
    colonIdx = prev;
  }
  return colonIdx;
}

export function stripSubagentSpinnerPrefix(text: string): string {
  return text.trim().replace(SUBAGENT_SPINNER_PREFIX, "").trim();
}

/** Runtime `pendingAuxState()` for main-thread thinking/compressing (not subagent status lines). */
export function isLivePendingReasoningAux(pendingAux: PendingAssistantAux | undefined): boolean {
  return Boolean(
    pendingAux &&
    (pendingAux.kind === "thinking" || pendingAux.kind === "compressing") &&
    !parsePendingSubagentStatusText(pendingAux.statusText),
  );
}

/** Runtime `pendingAuxState().statusText` while no `detailText` yet (e.g. `| Thinking…`). */
export function isGenericPendingThinkingStatusText(text: string | undefined): boolean {
  const normalized = text?.trim();
  if (!normalized) {
    return false;
  }
  const withoutSpinner = stripSubagentSpinnerPrefix(normalized);
  return withoutSpinner === "Thinking…";
}

/** Placeholder compaction aux before summary text arrives (e.g. `| Compressing…`). */
export function isGenericPendingCompactionStatusText(text: string | undefined): boolean {
  const normalized = text?.trim();
  if (!normalized) {
    return false;
  }
  const withoutSpinner = stripSubagentSpinnerPrefix(normalized);
  return withoutSpinner === "Compressing…";
}

function isSubagentRuntimeStatusTail(after: string): boolean {
  const tail = after.trim();
  if (!tail) {
    return false;
  }
  if (SUBAGENT_STATUS_TAIL_PREFIX.test(tail)) {
    return true;
  }
  if (/^The user wants\b/u.test(tail)) {
    return true;
  }
  if (/^Running\s*$/u.test(tail)) {
    return true;
  }
  if (tail.startsWith("Awaiting")) {
    return true;
  }
  if (/^Done\s*$/u.test(tail)) {
    return true;
  }
  return false;
}

/**
 * Runtime status like `task: Running` or `task: The user wants…` — not child final output
 * or parent post-tool summary (Markdown), and not normal assistant prose that happens to
 * contain a colon.
 */
export function isSubagentStatusSurfaceText(text: string | undefined): boolean {
  const normalized = text?.trim();
  if (!normalized) {
    return false;
  }
  if (isParentSubagentCompletionSurfaceText(normalized)) {
    return false;
  }
  if (/\*\*[\s\S]+\*\*/u.test(normalized) || /^#{1,6}\s/m.test(normalized)) {
    return false;
  }
  // Assistant replies are often multi-line; subagent status is always a single line.
  if (/[\r\n]/.test(normalized)) {
    return false;
  }

  const withoutSpinner = stripSubagentSpinnerPrefix(normalized);

  if (withoutSpinner === "Thinking…" || withoutSpinner === "Compressing…") {
    return true;
  }
  if (/:\s*Running\s*$/u.test(withoutSpinner)) {
    return true;
  }
  if (/:\s*Awaiting/u.test(withoutSpinner)) {
    return true;
  }

  const colonIdx = lastStatusColonIndex(withoutSpinner);
  if (colonIdx <= 0) {
    return false;
  }

  const before = withoutSpinner.slice(0, colonIdx).trim();
  const after = withoutSpinner.slice(colonIdx + 1).trim();
  if (!after || before.length < 4 || after.startsWith("```")) {
    return false;
  }

  if (isParentSubagentCompletionSurfaceText(before)) {
    return false;
  }

  // Runtime truncates progress to ~120 chars; real answers are longer.
  if (withoutSpinner.length > 220) {
    return false;
  }

  // Prose after a colon (lists, sentences) is not subagent progress.
  if (/[。！？；*•]/.test(after) || /^\s*[-*]\s/m.test(after)) {
    return false;
  }

  if (!isSubagentRuntimeStatusTail(after)) {
    return false;
  }

  if (after.length > 200) {
    return false;
  }

  return true;
}

export function hasSubagentToolInCurrentTurn(
  messages: ReadonlyArray<ConversationMessageSnapshot>,
): boolean {
  let lastUserIndex = -1;
  for (let index = messages.length - 1; index >= 0; index -= 1) {
    if (messages[index]?.role === "user") {
      lastUserIndex = index;
      break;
    }
  }
  for (let index = lastUserIndex + 1; index < messages.length; index += 1) {
    const message = messages[index];
    if (message?.role === "assistant" && message.tool?.toolName === "subagent") {
      return true;
    }
  }
  return false;
}

export function hasActiveSubagentToolInMessages(
  messages: ReadonlyArray<ConversationMessageSnapshot>,
): boolean {
  return messages.some(
    (message) =>
      message.role === "assistant" &&
      message.tool?.toolName === "subagent" &&
      (message.tool.phase === "preview" || message.tool.phase === "running"),
  );
}

/** While a subagent is still executing or awaiting approval in the current turn, the main timeline must not show child-session tool cards. */
export function hasInFlightSubagentDelegationInMessages(
  messages: ReadonlyArray<ConversationMessageSnapshot>,
): boolean {
  return messages.some(
    (message) =>
      message.role === "assistant" &&
      message.tool?.toolName === "subagent" &&
      (message.tool.phase === "preview" ||
        message.tool.phase === "running" ||
        message.tool.phase === "pending-approval"),
  );
}

export function isSubagentStatusSurfaceMessage(
  message: ConversationMessageSnapshot | undefined,
): boolean {
  return Boolean(
    message?.role === "assistant" && !message.tool && isSubagentStatusSurfaceText(message.content),
  );
}

/** Parsed subagent runtime status from `pendingAuxState().statusText` (spinner stripped). */
export function parsePendingSubagentStatusText(text: string | undefined): string | undefined {
  if (!text) {
    return undefined;
  }

  const status = stripSubagentSpinnerPrefix(text);
  if (!status || status === "Thinking…" || status === "Compressing…") {
    return undefined;
  }

  if (!isSubagentStatusSurfaceText(status)) {
    return undefined;
  }

  return status;
}

File: apps/cli/src/subagent_display.rs
use crate::view::{ChatMessage, ToolUiPhase};

const SUBAGENT_SPINNER_PREFIXES: [&str; 4] = ["| ", "/ ", "- ", "\\ "];

pub fn strip_subagent_spinner_prefix(text: &str) -> String {
    let trimmed = text.trim();
    for prefix in SUBAGENT_SPINNER_PREFIXES {
        if let Some(rest) = trimmed.strip_prefix(prefix) {
            return rest.trim().to_string();
        }
    }
    trimmed.to_string()
}

fn is_emoticon_colon(text: &str, colon_idx: usize) -> bool {
    text.as_bytes()
        .get(colon_idx + 1)
        .is_some_and(|next| b")D(P/\\]oO0-3".contains(next))
}

fn last_status_colon_index(text: &str) -> Option<usize> {
    let mut colon_idx = text
        .rfind(':')
        .into_iter()
        .chain(text.rfind('：'))
        .max()
        .unwrap_or(usize::MAX);
    if colon_idx == usize::MAX {
        return None;
    }
    while colon_idx > 0 && is_emoticon_colon(text, colon_idx) {
        let prev_ascii = text[..colon_idx].rfind(':');
        let prev_full = text[..colon_idx].rfind('：');
        colon_idx = prev_ascii.into_iter().chain(prev_full).max().unwrap_or(0);
        if colon_idx == 0 {
            return None;
        }
    }
    Some(colon_idx)
}

fn is_subagent_runtime_status_tail(after: &str) -> bool {
    let tail = after.trim();
    if tail.is_empty() {
        return false;
    }
    if tail.starts_with("The")
        || tail.starts_with("Sub")
        || tail.starts_with("Sp")
        || tail.starts_with("Thinking")
        || tail.starts_with("Compacting")
        || tail.starts_with("Running")
        || tail.starts_with("Awaiting")
        || tail.starts_with("正在")
    {
        return true;
    }
    if tail.starts_with("The user wants") || tail.starts_with("The user is") {
        return true;
    }
    if matches!(tail, "Running" | "Done" | "成功" | "完成") {
        return true;
    }
    if tail.starts_with("Awaiting") {
        return true;
    }
    // Short CJK-only progress fragments (e.g. "成功", "正在执行") — not English reasoning.
    tail.chars().count() <= 16 && !tail.chars().any(|ch| ch.is_ascii_alphabetic())
}

fn is_parent_subagent_completion_surface_text(text: &str) -> bool {
    text.contains("子智能体已完成") || text.contains("输出如下")
}

/// Mirrors Desktop `isSubagentStatusSurfaceText` for runtime status lines.
pub fn is_subagent_status_surface_text(text: &str) -> bool {
    let normalized = text.trim();
    if normalized.is_empty() {
        return false;
    }
    if is_parent_subagent_completion_surface_text(normalized) {
        return false;
    }
    if normalized.contains("**") || normalized.contains("\n#") {
        return false;
    }
    if normalized.contains('\n') || normalized.contains('\r') {
        return false;
    }

    let without_spinner = strip_subagent_spinner_prefix(normalized);
    if without_spinner == "Thinking…" || without_spinner == "Compacting…" {
        return true;
    }
    if without_spinner.ends_with(": Running") || without_spinner.ends_with("： Running") {
        return true;
    }
    if without_spinner.contains(": Awaiting") || without_spinner.contains("： Awaiting") {
        return true;
    }

    let colon_idx = match last_status_colon_index(&without_spinner) {
        Some(idx) if idx > 0 => idx,
        _ => return false,
    };
    let before = without_spinner[..colon_idx].trim();
    let after = if let Some(rest) = without_spinner[colon_idx..].strip_prefix(':') {
        rest.trim()
    } else if let Some(rest) = without_spinner[colon_idx..].strip_prefix('：') {
        rest.trim()
    } else {
        return false;
    };

    if after.is_empty() || before.len() < 4 || after.starts_with("```") {
        return false;
    }
    if is_parent_subagent_completion_surface_text(before) {
        return false;
    }
    if without_spinner.chars().count() > 220 {
        return false;
    }
    if after.contains('。')
        || after.contains('！')
        || after.contains('？')
        || after.contains('；')
        || after.contains('*')
        || after.contains('•')
        || after.contains(". ")
        || after.contains("?")
        || after.contains('!')
    {
        return false;
    }
    if !is_subagent_runtime_status_tail(after) {
        return false;
    }
    after.chars().count() <= 200
}

pub fn parse_pending_subagent_status_text(text: &str) -> Option<String> {
    let status = strip_subagent_spinner_prefix(text);
    if status.is_empty() || status == "Thinking…" || status == "Compacting…" {
        return None;
    }
    if !is_subagent_status_surface_text(&status) {
        return None;
    }
    Some(status)
}

pub fn has_active_subagent_tool_in_messages(messages: &[ChatMessage]) -> bool {
    messages.iter().any(|message| {
        message.tool_block.as_ref().is_some_and(|tool| {
            tool.tool_name == "subagent"
                && matches!(
                    tool.phase,
                    ToolUiPhase::Preview | ToolUiPhase::Running | ToolUiPhase::PendingApproval
                )
        })
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::view::{MessageRole, ToolUiBlock};

    #[test]
    fn detects_runtime_subagent_status_lines() {
        assert!(is_subagent_status_surface_text(
            "输出 \"你好\" 这两个字，不要做任何其他事情。: Running"
        ));
        assert!(is_subagent_status_surface_text(
            "请输出\"你好\"这两个字。: The"
        ));
        assert!(!is_subagent_status_surface_text("你好"));
    }

    #[test]
    fn rejects_prose_with_sentence_punctuation_after_colon() {
        assert!(!is_subagent_status_surface_text(
            "输出\"你好\"两个字。: The user is asking me to output \"你好\" (two Chinese characters meaning \"hello\"). This is a very simple and straightforward request."
        ));
    }

    #[test]
    fn active_subagent_tool_hides_standalone_status_surface() {
        let messages = vec![ChatMessage::with_tool_block(
            MessageRole::Agent,
            String::new(),
            ToolUiBlock {
                tool_call_id: Some("tc-1".to_string()),
                tool_name: "subagent".to_string(),
                phase: ToolUiPhase::Running,
                headline: "SubAgent".to_string(),
                detail_lines: vec![],
                image_paths: vec![],
                video_paths: vec![],
                args_excerpt: None,
                output_excerpt: None,
                suppress_expand: None,
            },
        )];
        assert!(has_active_subagent_tool_in_messages(&messages));
    }
}
